share/functions/edit_command_buffer.fish

```diff
@@ -0,0 +1,45 @@
+function edit_command_buffer --description 'Edit the command buffer in an external editor'
+    set -l f (mktemp)
+    if set -q f[1]
+        mv $f $f.fish
+        set f $f.fish
+    else
+        # We should never execute this block but better to be paranoid.
+        if set -q TMPDIR
+            set f $TMPDIR/fish.(echo %self).fish
+        else
+            set f /tmp/fish.(echo %self).fish
+        end
+        touch $f
+        or return 1
+    end
+
+    # Edit the command line with the users preferred editor or vim or emacs.
+    commandline -b >$f
+    if set -q EDITOR
+        eval $EDITOR $f
+    else if command -s vim >/dev/null
+        vim $f
+    else if command -s emacs >/dev/null
+        emacs $f
+    else
+        echo
+        echo (_ 'Sorry, but you did not define $EDITOR and I could not find vim or emacs.')
```

Comment:
In general, I think we should be trying for non-first-person error messages.

How about: "External editor requested but $EDITOR / $VISUAL not set, vim or emacs not available."
"Try setting the EDITOR variable to a text editor, or installing vim or emacs."